### PR TITLE
Fix/3689/navigate on delete thread

### PR DESCRIPTION
**Pull request by StCyr**
fix to https://github.com/nextcloud/mail/issues/3689

Please review

### CODE REVIEW

**Comment by StCyr** [comment]
oups, there's a commit that shouldn't be there... hold on

**Comment by StCyr** [comment]
> oups, there's a commit that shouldn't be there... hold on

Ok resolved. sorry

File: src/components/Thread.vue [hunk 1]
```diff
@@ -50,26 +51,23 @@ export default {
 			message: undefined,
 			errorMessage: '',
 			error: undefined,
+			thread: [],
```
Comment (ChristophWurst) [inline src/components/Thread.vue:54]: :-1: this makes the thread non-reactive

File: src/components/Thread.vue [hunk 2]
```diff
@@ -50,26 +51,23 @@ export default {
...
-			return this.$store.getters.getEnvelopeThread(this.$route.params.threadId)
-		},
 		threadParticipants() {
 			const recipients = this.thread.flatMap(envelope => {
 				return envelope.from.concat(envelope.to).concat(envelope.cc)
 			})
 			return uniqBy(prop('email'), recipients)
 		},
 		threadSubject() {
-			const thread = this.thread
```
Comment (ChristophWurst) [inline src/components/Thread.vue:67]: unrelated change?

**Review by ChristophWurst — CHANGES_REQUESTED** [review]
We should update the thread data in the store than mangling with it in a local copy. If you use this solution you'll run into the same issue if you navigate somewhere else and back: the thread change is undone.

File: src/components/Thread.vue (hunk 2 above)
Comment (StCyr) [reply src/components/Thread.vue:67]: yes, unrelated change. 
just, I think this line doesn't serve any purpose...?

File: src/components/Thread.vue (hunk 1 above)
Comment (StCyr) [reply src/components/Thread.vue]: new version proposed. Shall I create a new PR?

**Comment by StCyr** [comment]
I think I've figured out how thread info were to be managed.

Could you have a look at this PR again @ChristophWurst ?

**Review by ChristophWurst — APPROVED** [review]
So right now when deleting a message you

* fetch the thread of this message (1 additional request)
* fetch the thread of all messages in this thread (n additional requests)
* dispatch n+1 mutations

Can't we do it a bit more optimistic and either just update all known envelopes from a message's `thread` property or all other messages with the same `mailboxId` and `threadRootId`?

**Comment by StCyr** [comment]
> 
> 
> So right now when deleting a message you
> 
>     * fetch the thread of this message (1 additional request)
> 
>     * fetch the thread of all messages in this thread (n additional requests)
> 
>     * dispatch n+1 mutations
 
Yes, I agree that may be a lot of requests.

> 
> Can't we do it a bit more optimistic and either just update all known envelopes from a message's `thread` property or all other messages with the same `mailboxId` and `threadRootId`?

Wouldn't it be better that the backend provides the thread property directly on a fetchEnvelope call?

**Comment by ChristophWurst** [comment]
> Wouldn't it be better that the backend provides the thread property directly on a fetchEnvelope call?

Ideally, yes. But we have to be careful as this easily increases the runtime complexity from O(1) to O(n+1) if we have to load the thread in one query at a time for every message. The relational database won't allow us to have the list of messages with the threads in one query.

**Comment by ChristophWurst** [comment]
@StCyr conflicts and no update in a while – are you still working on this?

**Comment by StCyr** [comment]
No. You may close it if you like
